File: admin-web/src/lib/api/technician-jobs.js
import apiClient from "$lib/api/http";

function normalizeJob(raw = {}) {
  return {
    id: raw.id ?? null,
    hubId: raw.hub_id ?? null,
    hubCode: raw.hub_code ?? "",
    hubName: raw.hub_name ?? "",
    hardwareId: raw.hardware_id ?? null,
    hardwareImei: raw.hardware_imei ?? "",
    hardwareModel: raw.hardware_model ?? "",
    vehicleId: raw.vehicle_id ?? null,
    assignedTechnicianId: raw.assigned_technician_id ?? null,
    assignedTechnicianName: raw.assigned_technician_name ?? "",
    assignedTechnicianEmail: raw.assigned_technician_email ?? "",
    requestedById: raw.requested_by_id ?? null,
    requestedByName: raw.requested_by_name ?? "",
    status: raw.status ?? "pending",
    priority: raw.priority ?? "normal",
    scheduledFor: raw.scheduled_for ?? null,
    startedAt: raw.started_at ?? null,
    acceptedAt: raw.accepted_at ?? null,
    completedAt: raw.completed_at ?? null,
    cancelledAt: raw.cancelled_at ?? null,
    declinedAt: raw.declined_at ?? null,
    installedAt: raw.installed_at ?? null,
    installationLocation: raw.installation_location ?? "",
    installationLatitude: raw.installation_latitude ?? null,
    installationLongitude: raw.installation_longitude ?? null,
    assetLabel: raw.asset_label ?? "",
    assetRegistration: raw.asset_registration ?? "",
    notes: raw.notes ?? "",
    completionNotes: raw.completion_notes ?? "",
    declineReason: raw.decline_reason ?? "",
    assignmentReference: raw.assignment_reference ?? "",
    createdAt: raw.created_at ?? null,
    updatedAt: raw.updated_at ?? null,
  };
}

export async function fetchTechnicianJobs(params = {}) {
  const { data } = await apiClient.get("/technician-jobs", { params });
  const items = Array.isArray(data?.items) ? data.items.map(normalizeJob) : [];
  const meta = data?.meta ?? { page: 1, per_page: items.length, total: items.length };
  return { items, meta };
}

export async function fetchTechnicians() {
  const { data } = await apiClient.get("/users/technicians");
  return Array.isArray(data)
    ? data.map((item) => ({
        id: item.id ?? null,
        name: item.name ?? "",
        email: item.email ?? "",
        role: item.role ?? "technician",
      }))
    : [];
}

export async function createTechnicianJob(payload = {}) {
  const requestBody = {
    hub_id: payload.hubId,
    hardware_id: payload.hardwareId ? Number(payload.hardwareId) : undefined,
    vehicle_id: payload.vehicleId ?? undefined,
    assigned_technician_id: payload.assignedTechnicianId ?? undefined,
    priority: payload.priority ?? "normal",
    scheduled_for: payload.scheduledFor ?? undefined,
    installed_at: payload.installedAt ?? undefined,
    installation_location: payload.installationLocation ?? undefined,
    installation_latitude: payload.installationLatitude ?? undefined,
    installation_longitude: payload.installationLongitude ?? undefined,
    asset_label: payload.assetLabel ?? undefined,
    asset_registration: payload.assetRegistration ?? undefined,
    notes: payload.notes ?? undefined,
  };
  const { data } = await apiClient.post("/technician-jobs", requestBody);
  return normalizeJob(data ?? requestBody);
}

export async function updateTechnicianJob(jobId, payload = {}) {
  if (!jobId) {
    throw new Error("jobId is required");
  }
  const requestBody = {
    status: payload.status ?? undefined,
    assigned_technician_id: payload.assignedTechnicianId ?? undefined,
    priority: payload.priority ?? undefined,
    scheduled_for: payload.scheduledFor ?? undefined,
    installed_at: payload.installedAt ?? undefined,
    installation_location: payload.installationLocation ?? undefined,
    installation_latitude: payload.installationLatitude ?? undefined,
    installation_longitude: payload.installationLongitude ?? undefined,
    asset_label: payload.assetLabel ?? undefined,
    asset_registration: payload.assetRegistration ?? undefined,
    notes: payload.notes ?? undefined,
    completion_notes: payload.completionNotes ?? undefined,
    decline_reason: payload.declineReason ?? undefined,
  };
  const { data } = await apiClient.patch(`/technician-jobs/${jobId}`, requestBody);
  return normalizeJob(data ?? { id: jobId, ...requestBody });
}

File: backend/app/schemas/technician_jobs.py
"""Schemas for technician workflow board APIs."""
from __future__ import annotations

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.technician_job import TechnicianJobPriority, TechnicianJobStatus


class TechnicianJobItem(BaseModel):
    """Serialized technician job card."""

    id: UUID
    hub_id: UUID
    hub_code: Optional[str] = None
    hub_name: Optional[str] = None
    hardware_id: Optional[int] = None
    hardware_imei: Optional[str] = None
    hardware_model: Optional[str] = None
    vehicle_id: Optional[UUID] = None
    assigned_technician_id: Optional[UUID] = None
    assigned_technician_name: Optional[str] = None
    assigned_technician_email: Optional[str] = None
    requested_by_id: Optional[UUID] = None
    requested_by_name: Optional[str] = None
    status: TechnicianJobStatus
    priority: TechnicianJobPriority
    scheduled_for: Optional[datetime] = None
    started_at: Optional[datetime] = None
    accepted_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    cancelled_at: Optional[datetime] = None
    declined_at: Optional[datetime] = None
    installed_at: Optional[datetime] = None
    installation_location: Optional[str] = None
    installation_latitude: Optional[float] = None
    installation_longitude: Optional[float] = None
    asset_label: Optional[str] = None
    asset_registration: Optional[str] = None
    notes: Optional[str] = None
    completion_notes: Optional[str] = None
    decline_reason: Optional[str] = None
    assignment_reference: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class TechnicianJobsListResponse(BaseModel):
    """Paginated list response."""

    items: list[TechnicianJobItem]
    meta: dict


class TechnicianJobCreateRequest(BaseModel):
    """Admin payload for creating a job card."""

    hub_id: UUID
    hardware_id: Optional[int] = None
    vehicle_id: Optional[UUID] = None
    assigned_technician_id: Optional[UUID] = None
    priority: TechnicianJobPriority = TechnicianJobPriority.normal
    scheduled_for: Optional[datetime] = None
    installed_at: Optional[datetime] = None
    installation_location: Optional[str] = Field(default=None, max_length=255)
    installation_latitude: Optional[float] = None
    installation_longitude: Optional[float] = None
    asset_label: Optional[str] = Field(default=None, max_length=255)
    asset_registration: Optional[str] = Field(default=None, max_length=100)
    notes: Optional[str] = None


class TechnicianJobUpdateRequest(BaseModel):
    """Patch payload for admin/technician actions."""

    status: Optional[TechnicianJobStatus] = None
    assigned_technician_id: Optional[UUID] = None
    priority: Optional[TechnicianJobPriority] = None
    scheduled_for: Optional[datetime] = None
    installed_at: Optional[datetime] = None
    installation_location: Optional[str] = Field(default=None, max_length=255)
    installation_latitude: Optional[float] = None
    installation_longitude: Optional[float] = None
    asset_label: Optional[str] = Field(default=None, max_length=255)
    asset_registration: Optional[str] = Field(default=None, max_length=100)
    notes: Optional[str] = None
    completion_notes: Optional[str] = None
    decline_reason: Optional[str] = None
